ports/stm32/ulpi.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2016 STMicroelectronics. All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ * USB3320 ULPI functions ported from stm32f7xx_lp_modes.c
+ */
+#include <stdint.h>
+#include "py/mphal.h"
+#include "pin_static_af.h"
+#include "ulpi.h"
+
+#if (MICROPY_HW_USB_HS_ULPI3320 == 1)
+#include STM32_HAL_H
+
+#define USBULPI_PHYCR       ((uint32_t)(0x40040000 + 0x034))
+#define USBULPI_D07         ((uint32_t)0x000000FF)
+#define USBULPI_New         ((uint32_t)0x02000000)
+#define USBULPI_RW          ((uint32_t)0x00400000)
+#define USBULPI_S_BUSY      ((uint32_t)0x04000000)
+#define USBULPI_S_DONE      ((uint32_t)0x08000000)
+
+#define USB_OTG_READ_REG32(reg)  (*(__IO uint32_t *)(reg))
+#define USB_OTG_WRITE_REG32(reg,value) (*(__IO uint32_t *)(reg) = (value))
+
+extern void __fatal_error(const char *msg);
+
+/**
+  * @brief  Read CR value
+  * @param  Addr the Address of the ULPI Register
+  * @retval Returns value of PHY CR register
+  */
+static uint32_t USB_ULPI_Read(uint32_t Addr) {
+    __IO uint32_t val = 0;
+    __IO uint32_t timeout = 100; /* Can be tuned based on the Clock or etc... */
```

Comment:
Do these variables really need to be `__IO` (volatile)?  It looks like the code will work OK if they don't have `__IO`.

Also might be worth making the timeout value a constant in this file, like `ULPI_READ_TIMEOUT_COUNT` (it's not measured in milliseconds or anything, just a count).